pkg/policyStatus/0_main.go

```diff
@@ -0,0 +1,87 @@
+package policyStatus
+
+import (
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	"github.com/nirmata/kyverno/pkg/policystore"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/nirmata/kyverno/pkg/client/clientset/versioned"
+
+	v1 "github.com/nirmata/kyverno/pkg/api/kyverno/v1"
+)
+
+type Sync struct {
+	cache       *cache
+	listener    chan statusUpdater
+	stop        <-chan struct{}
+	client      *versioned.Clientset
+	policyStore *policystore.PolicyStore
+}
+
+type cache struct {
+	mutex sync.RWMutex
+	data  map[string]v1.PolicyStatus
+}
+
+func NewSync(c *versioned.Clientset, sc <-chan struct{}, pms *policystore.PolicyStore) *Sync {
+	return &Sync{
+		cache: &cache{
+			mutex: sync.RWMutex{},
+			data:  make(map[string]v1.PolicyStatus),
+		},
+		stop:        sc,
+		client:      c,
+		policyStore: pms,
+		listener:    make(chan statusUpdater),
+	}
+}
+
+func (s *Sync) Run(workers int) {
+	for i := 0; i < workers; i++ {
+		go s.updateStatusCache()
+	}
+
+	wait.Until(s.updatePolicyStatus, 2*time.Second, s.stop)
+	<-s.stop
+	s.updatePolicyStatus()
```

Comment:
It is cleaning up the status cache between to updates, and not accepting new status update request. Since we have implemented graceful shutdown, we should enable these kind of features - otherwise there is no difference between graceful shutdown and force killing the task.

Reply:
From implementation point of view, we only react to signals(SIGTERM & SIGINT) for stopCh. With `updatePolicyStatus` we are updating the status from then channel. My concern was this is a best-effort scenario, so if the interrupt is received and we have time to cleanUp then we will capture the stats. But if we receive an interrupt and don't have time to cleanUp the end result will be different, this is more like a post-shutdown step. But in the interest of time, we can let this be and reevaluate if needed.

Reply:
requested changes have been made